src/coreclr/jit/optcse.cpp

```diff
@@ -478,8 +478,9 @@ unsigned Compiler::optValnumCSE_Index(GenTree* tree, Statement* stmt)
         assert(vnStore->IsVNConstant(vnLibNorm));
 
         // We don't share small offset constants when they require a reloc
+        // Also, we don't share non-null const gc handles
         //
-        if (!tree->AsIntConCommon()->ImmedValNeedsReloc(this))
+        if (!tree->AsIntConCommon()->ImmedValNeedsReloc(this) && ((tree->IsIntegralConst(0)) || !tree->TypeIs(TYP_REF)))
```

Comment:
Should we use `varTypeIsGC` instead? I imagine we can have the same problem with `TYP_BYREF`. Or is producing `TYP_BYREF` via pointer arithmetic of a `TYP_BYREF` that may be pointing into another object ok in this particular case?

Reply:
Yeah, let's use `varTypeIsGC`